client.go

```diff
@@ -118,59 +133,167 @@ type RequestParams struct {
 	Response any
 }
 
-// Generic helper to execute HTTP requests using the
+// Generic helper to execute HTTP requests using the net/http package
 //
-//nolint:unused
+// nolint:unused
 func (c *httpClient) doRequest(ctx context.Context, method, url string, params RequestParams, mutators ...func(req *http.Request) error) error {
-	// Create a new HTTP request
+	req, bodyBuffer, err := c.createRequest(ctx, method, url, params)
+	if err != nil {
+		return err
+	}
+
+	if err := c.applyMutators(req, mutators); err != nil {
+		return err
+	}
+
+	if c.debug && c.logger != nil {
+		c.logRequest(req, method, url, bodyBuffer)
+	}
+
+	resp, err := c.sendRequest(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if err := c.checkHTTPError(resp); err != nil {
+		return err
+	}
+
+	if c.debug && c.logger != nil {
+		resp, err = c.logResponse(resp)
+		if err != nil {
+			return err
+		}
+	}
+
+	if params.Response != nil {
+		if err := c.decodeResponseBody(resp, params.Response); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// nolint:unused
+func (c *httpClient) createRequest(ctx context.Context, method, url string, params RequestParams) (*http.Request, *bytes.Buffer, error) {
 	var bodyReader io.Reader
+	var bodyBuffer *bytes.Buffer
+
 	if params.Body != nil {
-		buf := new(bytes.Buffer)
-		if err := json.NewEncoder(buf).Encode(params.Body); err != nil {
-			return fmt.Errorf("failed to encode body: %w", err)
+		bodyBuffer = new(bytes.Buffer)
+		if err := json.NewEncoder(bodyBuffer).Encode(params.Body); err != nil {
+			if c.debug && c.logger != nil {
+				c.logger.Errorf("failed to encode body: %v", err)
+			}
+			return nil, nil, fmt.Errorf("failed to encode body: %w", err)
 		}
-		bodyReader = buf
+		bodyReader = bodyBuffer
 	}
 
 	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
 	if err != nil {
-		return fmt.Errorf("failed to create request: %w", err)
+		if c.debug && c.logger != nil {
+			c.logger.Errorf("failed to create request: %v", err)
+		}
+		return nil, nil, fmt.Errorf("failed to create request: %w", err)
 	}
 
-	// Set default headers
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("Accept", "application/json")
 	if c.userAgent != "" {
 		req.Header.Set("User-Agent", c.userAgent)
 	}
 
-	// Apply mutators
+	return req, bodyBuffer, nil
+}
+
+// nolint:unused
+func (c *httpClient) applyMutators(req *http.Request, mutators []func(req *http.Request) error) error {
 	for _, mutate := range mutators {
 		if err := mutate(req); err != nil {
+			if c.debug && c.logger != nil {
+				c.logger.Errorf("failed to mutate request: %v", err)
+			}
 			return fmt.Errorf("failed to mutate request: %w", err)
 		}
 	}
+	return nil
+}
+
+// nolint:unused
+func (c *httpClient) logRequest(req *http.Request, method, url string, bodyBuffer *bytes.Buffer) {
+	var reqBody string
+	if bodyBuffer != nil {
+		reqBody = bodyBuffer.String()
```

Comment:
nitpick: Could we display `nil` if the given body is nil here just to make the output a tiny bit more precise?

Reply:
Fixed! Thanks for the tip.